### ISSUE TITLE
Enable option to activate the control port on 6082

### ISSUE THREAD

**Issue by arthurlogilab** [IBM/varnish-operator#5]
First off thanks for varnish-operator, we're very happy with the way it works and are using it on one of our projects. 

The application it is caching has a mechanism which invalidates URLs using the control port (usually on 6082) with a shared secret to authenticate. 

By default the operator puts `-T 127.0.0.1:6082`, so we found a way round this with `args`, and setting the secret with `admAuth` : 

```yaml
---
apiVersion: caching.ibm.com/v1alpha1
kind: VarnishCluster
metadata:
  labels:
    operator: varnish
  name: varnish-cluster
spec:
   [snip]
    admAuth:
        secretName: varnish-custom-secret
        key: varnish-custom-secret
    args: ["-T", "0.0.0.0:6082"]
```
We also need to add a Service object that points to that so it can be acceded from the pods using `varnish-cli:6082`. 

Could there be a less "manual" way to make it easier for other people ? We'd be happy to give feedback and help test if there is a way to enable this type of feature.

**Comment by arthurlogilab** [comment]
We're a bit stuck actually, do you know if there is a way to patch the Pods/StatefullSets to overrid the open ports and the command line options ? It seems that whenever we change something the operator "fixes" the state by changing it back.

**Comment by cin** [comment]
Thanks @arthurlogilab . Sorry for the delay. We're actually in meetings at the moment but will discuss ASAP.  The port issue will have to be addressed in the operator's code as it will always revert your changes. You should be able to alter the command line options (just not by editing the pod/sts). @tomashibm for visibility.

**Comment by cin** [comment]
We are going to work on a fix for this but have a work around for you to try in the meantime. This is less than ideal but you can stop the operator so it won't overwrite your changes. We also validated that the loopback address being is being maintained by the operator for the control port (that's why your command line changes weren't sticking).

- Stop the operator so it won't revert your changes
- Modify the sts
  - Change the sts' command to 0.0.0.0 as you did above
  - Expose 6082 in the sts definition
- Either manually create a new service to point to 6082 or modify the existing service to expose 6082

I'm assuming you're running varnish in a non-sharded mode (i.e. one instance of varnish)? I'm not sure how the service should behave in a sharded setup (i.e. we have some setups that use 32 varnish instances). The service would essentially round-robin the requests you want to make to the various varnish pods. So we need to think a bit on how to handle this scenario.

**Comment by arthurlogilab** [comment]
Hey @cin thanks for taking the time to respond, you don't have to apologize for delays, this is an open source project, we're already grateful for the work you've done, we don't expect round the clock support or feedback. So thanks again. 

About the workaround, I'm a bit new to operators, I can't seem to find a translation to "Stop the operator so it won't revert your changes", I'll continue search, but if you have an answer, I'd welcome it. 

Yes we are running a "simple" setup, no sharding. Indeed the sharding setup can be more tricky for purging objects, so let's say it's not in the scope of this issue.

**Comment by arthurlogilab** [comment]
Haven't found any documentation of "how to stop an operator", so I'm going with the following hack : 

* download statefulset
* delete VarnishCluster operator
* cleanup statefullset yaml
* adapt for the port and -T option
* kubectl apply -f patched-statefulset.yaml

**Comment by tomashibm** [comment]
yeap, you were on the right way. We, instead of deleting, just scale down the deployment to zero. Has the same effect.

Could you expand more on how you use the control port to invalidate URLs? we'd like to understand better the use case.

The issue that might be there is that with multiple instances, your app will connect and invalidate the URL only on one node (whichever it is round-robined to). Not sure if Varnish is able coordinate that kind of operations so that it gets applied to all instances.

**Comment by arthurlogilab** [comment]
Our usecase is the following : the web application knows it's behind a varnish cache, so when an content editor modifies example.org/article/123 the web application, connects to varnish using varnishcli on port 6082 (configurable) and sends a `ban url.req == /article/123` to it (simplifying here, since in some cases multiple URLs can be purged/ban when a content changes). This way we can have a very high TTL and serve the (mostly low-rate-of-change) site at much better speeds. 

As for the invalidate cache on multiple varnish, the web application in question can have a list of varnishcli hosts to talk to, so we might have a way to configure that if the varnish servers don't talk to each other.

**Comment by arthurlogilab** [comment]
We though we solved the problem and bypassed the bug, but only part of it, the patched stateful set only solves the "making port 8062 accessible" part but since `/etc/varnish/default.vcl` is no longer generated by the operator the varnish service is useless. (no config, no backends, no pages served). We tried re-adding the services but that wasn't enough, there is some piece missing.

**Comment by cin** [comment]
Ah, that makes sense. We didn't consider the VCL generation when thinking of this workaround.

Thanks for the use case update. Is this something you run with a curl command? If not, how are you making the ban call? An example would be nice here so we can test exactly what you're trying to do.

**Comment by tomashibm** [comment]
> As for the invalidate cache on multiple varnish, the web application in question can have a list of varnishcli hosts to talk to, so we might have a way to configure that if the varnish servers don't talk to each other.

yeah, that should work, but the missing piece for me here is the list of hosts. You have only the Kubernetes service as the URL to connect to Varnish. But sounds like you need the list of pod IPs so you know the address to each Varnish instance.

**Comment by tomashibm** [comment]
as for vcl generation with disabled operator:

From what you said @arthurlogilab looks like you first disabled the operator and then created the `VarnishCluster` resource, statefulset and service manually, right?
In that case - yes, there are a few other components that has to be deployed and configured correctly in order to setup the cluster.

But it still should be possible:
1. Have the operator deployed
2. Create your `VarnishCluster` and wait until it gets set up by the operator.
3. Scale the operator to 0 replicas by changing the deployment config.
4. Now you can make changes to the existing configured cluster and they won't get overwritten. Modify your statefulset and service as needed.

VCL won't get generated and loaded into your cluster if you deployed the cluster with the operator disabled. But once it is deployed by the operator, the VCL reload should happen without the operator. This is possible since the VCL generation and reload happens in the sidecar of the Varnish pod. But first it has to be configured by the operator. Let me know if it works for you.

**Comment by shveik** [comment]
@arthurlogilab instead of invalidating the cache through the api, maybe a different solution: 
1. create an internal api that's versioned and/or timestamped like: /internal/cms/v5/[02222021:12:34...]/myarticle.html
2. create another endpoint that returns that value like: GET /latest?article=request.params.articlename 
returns { name: myarticle.html, latest: /internal/cms/v5/[02222021:12:34...]/myarticle.html}
3. when request comes from client to varnish for the article like GET /myarticle.html, first, make the internal call, it should be very lightweight and cached for a short duration (one minute maybe if that's appropriate for your content?) and replace the url in the client call with the actual article value. 
that way, you don't have to mess with the control port and try to figure out where the cache is.

**Comment by arthurlogilab** [comment]
Thanks for your various feedbacks.  We ended up not finding any way to deploy without the operator, (that was before @tomashibm 's latest feedback, but I think we tried that and it didn't work). 

We're stuck, and our hosting partner has to go and do varnishctl commands by hand when needed. 

@shveik sorry to say that your suggestion feels like re-writing a cache application and feels out of the scope of the current bug report (but maybe I misunderstood) 

Has anyone taken a look at trying to fix the test added by https://github.com/IBM/varnish-operator/pull/7 ? I think we're past trying to find a workaround and should fix this in the operator.

**Comment by tomashibm** [comment]
the behavior in the test is intentional. We disallow certain arguments to be set since the operator relies on some configs to work properly. The control port is amongst them. So we override it if someone specifies it.
Will it work for you if we change it to be `0.0.0.0:6082`? I think it solves that part of the issue for you.

But still, the open question for me is how to go over each varnish node to do the `ban url.req == /url/123`. You said that your app can accept multiple Varnish hosts, but where would you get that list of hosts? The Kubernetes service URL won't work here if you have more than one node.

**Comment by tomashibm** [comment]
We made the management pod available outside of the pod in the release [0.27.2](https://github.com/IBM/varnish-operator/releases/tag/0.27.2). That should allow you to expose the management port and execute commands.
We didn't open that port on the service since with multiple pods it's unpredictable which pod you will execute the commands on which makes it not very useful in our opinion. 
If you need a service with that port opened, you can create it yourself having the `<varnishcluster-name>-cached` service as the reference.
Also you can hit the pod directly using the in-cluster DNS name.
Closing the issue, but feel free to reopen or create an another one if need any other help.

**Comment by arthurlogilab** [comment]
Thanks for doing this, we'll test this as soon as possible on our deployment !
